internal/mesh/internal/controllers/routes/controller.go

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package routes
+
+import (
+	"context"
+
+	"github.com/hashicorp/go-hclog"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	"github.com/hashicorp/consul/internal/catalog"
+	"github.com/hashicorp/consul/internal/controller"
+	"github.com/hashicorp/consul/internal/mesh/internal/controllers/routes/loader"
+	"github.com/hashicorp/consul/internal/mesh/internal/controllers/routes/xroutemapper"
+	"github.com/hashicorp/consul/internal/mesh/internal/types"
+	"github.com/hashicorp/consul/internal/resource"
+	pbmesh "github.com/hashicorp/consul/proto-public/pbmesh/v1alpha1"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+func Controller() controller.Controller {
+	mapper := xroutemapper.New()
+
+	r := &routesReconciler{
+		mapper: mapper,
+	}
+	return controller.ForType(types.ComputedRoutesType).
+		WithWatch(types.HTTPRouteType, mapper.MapHTTPRoute).
+		WithWatch(types.GRPCRouteType, mapper.MapGRPCRoute).
+		WithWatch(types.TCPRouteType, mapper.MapTCPRoute).
+		WithWatch(types.DestinationPolicyType, mapper.MapDestinationPolicy).
+		WithWatch(catalog.FailoverPolicyType, mapper.MapFailoverPolicy).
+		WithWatch(catalog.ServiceType, mapper.MapService).
+		WithReconciler(r)
+}
+
+type routesReconciler struct {
+	mapper *xroutemapper.Mapper
+}
+
+func (r *routesReconciler) Reconcile(ctx context.Context, rt controller.Runtime, req controller.Request) error {
+	// Notably don't inject the resource-id here into the logger, since we have
+	// to do a fan-out to multiple resources due to xRoutes having multiple
+	// parent refs.
+	rt.Logger = rt.Logger.With("controller", StatusKey)
+
+	rt.Logger.Trace("reconciling computed routes")
+
+	loggerFor := func(id *pbresource.ID) hclog.Logger {
+		return rt.Logger.With("resource-id", id)
+	}
+	related, err := loader.LoadResourcesForComputedRoutes(ctx, loggerFor, rt.Client, r.mapper, req.ID)
+	if err != nil {
+		rt.Logger.Error("error loading relevant resources", "error", err)
+		return err
+	}
+
+	pending := make(PendingStatuses)
+
+	ValidateXRouteReferences(related, pending)
+
+	generatedResults := GenerateComputedRoutes(related, pending)
+
+	if err := UpdatePendingStatuses(ctx, rt, pending); err != nil {
+		rt.Logger.Error("error updating statuses for affected relevant resources", "error", err)
+		return err
+	}
+
+	for _, result := range generatedResults {
+		computedRoutesID := result.ID
+
+		logger := rt.Logger.With("resource-id", computedRoutesID)
+
+		prev, err := resource.GetDecodedResource[*pbmesh.ComputedRoutes](ctx, rt.Client, computedRoutesID)
+		if err != nil {
+			logger.Error("error loading previous computed routes", "error", err)
+			return err
+		}
+
+		if err := ensureComputedRoutesIsSynced(ctx, logger, rt.Client, result, prev); err != nil {
+			return err
+		}
```

Comment:
If the server does it, it would only benefit the removal of some raft writes. If the client does it then you get the benefit of also avoiding sending the data over the network at all.